streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -299,18 +300,27 @@ public void resume() {
     }
 
     /**
-     * An active task is processable if its buffer contains data for all of its input source topic partitions
+     * An active task is processable if its buffer contains data for all of its input
+     * source topic partitions, or if it is enforced to be processable
      */
-    public boolean isProcessable() {
+    boolean isProcessable(final long now) {
         if (partitionGroup.allPartitionsBuffered()) {
+            idleStartTime = RecordQueue.NOT_KNOWN;
             return true;
-        } else if (partitionGroup.numBuffered() > 0 && --waits < 0) {
-            taskMetrics.taskEnforcedProcessSensor.record();
-            waits = WAIT_ON_PARTIAL_INPUT;
-            return true;
-        }
+        } else if (partitionGroup.numBuffered() > 0) {
+            if (idleStartTime == RecordQueue.NOT_KNOWN) {
+                idleStartTime = now;
+            }
 
-        return false;
+            if (now - idleStartTime >= maxTaskIdleMs) {
```

Comment:
There was an early comment on the test code that suggests `>=`. Personally I think it does not make a big difference at all.

Reply:
I am to blame for this suggestion. I agree it doesn't make a big difference.
The reasoning was that if it's the "maximum idle time", then you shouldn't idle longer than it, otherwise, it's not really a maximum.

Reply:
I agree that it does not matter too much :) (that why it's a nit)

However, I think that the maximum is inclusive, and only if we exceed it, we should force processing. From my understanding, "maximum idle time" is actually a lower bound (-> don't force processing until this time passed) because we cannot guarantee anyway to not exceed this threshold. I see your point why the name might be counter intuitive (even if I think the name is correct). If you interpret the name strictly, we would be allowed (or actually we would be required) to force processing before the time passed. This interpretation would make the parameter useless (ie, user tells us to idle max 5 minutes and we obey by forcing processing after 1 minute).

To me, the right interpretation is, "wait until this time passed and force processing asap if the time is exceeded". Chaning the name to `min.idle.time.ms` would be more precise, but I think it would be more confusing to users.

Reply:
Okay guys, I'm going to make a final call here to end the discussion: I'm staying with `max.idle..` since I feel it is easier to understand for users, and be aware that this is not strictly respected in practice unless it is set to `0`. Also I'm staying with `>=` since again, it is easier to understand though not strictly sound mathematically.

Reply:
:) Fair enough.